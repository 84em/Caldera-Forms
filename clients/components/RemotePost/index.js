import {RemotePost} from "./RemotePost";
import {FeaturedImage} from "./FeaturedImage";
import Image from "./Image";
import PostExcerpt from './PostExcerpt'
import PostTitle from './PostTitle'
import RemoteImage from './RemoteImage'
import PostCard from './PostCard'
export {
    RemotePost,
    FeaturedImage,
    Image,
    PostExcerpt,
    PostTitle,
    RemoteImage,
    PostCard
}

