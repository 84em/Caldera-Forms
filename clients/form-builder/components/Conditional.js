import React, {Fragment} from 'react';

/**
 * The section to set which fields it applies to
 *
 * @since 1.8.10
 *
 * @param Array formFields The form's fields
 * @param Array fieldsUsed The fields this conditional is applied to.
 * @param Array notAllowedFields The fields this conditional can NOT applied to.
 * @param strings Translation strings
 * @param onChange
 * @param groupId
 * @returns {*}
 * @constructor
 */
export const AppliesToFields = ({formFields, fieldsUsed, notAllowedFields, strings, onChange, groupId}) => {
    /**
     * Check if we should disable this option
     *
     * @since 1.8.10
     *
     * @param fieldId
     * @returns {*}
     */
    function isFieldDisabled(fieldId) {
        if (undefined === notAllowedFields) {
            return false;
        }
        return notAllowedFields.includes(fieldId);
    }

    /**
     * Check if field is used and should be checked.
     *
     * @since 1.8.10
     *
     * @param fieldId
     * @returns {*}
     */
    function isFieldChecked(fieldId) {
        return fieldsUsed.includes(fieldId);
    }

    return (
        <div style={{float: 'left', width: '288px', paddingLeft: '12px'}}>
            <h4 style={{borderBottom: `1px solid rgb(191, 191, 191)`, margin: `0px 0px 6px; padding: 0px 0px 6px`}}>
                {strings['applied-fields']}
            </h4>
            <p className="description">{strings['select-apply-fields']}</p>
            {formFields.map(field => (
                    <label style={{display: 'block', marginLeft: `20px`}}>
                        {field.label}
                        <input
                            style={{marginLeft: `-20px`}}
                            type="checkbox"
                            disabled={isFieldDisabled(field.ID)}
                            onClick={(e) => {
                                e.preventDefault();
                                //Is it already applied?
                                if (isFieldChecked(field.ID)) {
                                    //remove from applied fields
                                    onChange(fieldsUsed.filter(f => f.ID === field.ID))
                                } else {
                                    //Add to applied fields.
                                    onChange([...fieldsUsed, field.ID])
                                }
                            }}
                            value={groupId}
                            checked={isFieldChecked(field.ID)}
                        />
                    </label>
                )
            )}
        </div>
    );
};

const isFieldTypeWithOptions = (fieldType) => ['dropdown', 'checkbox', 'radio', 'filtered_select2', 'toggle_switch'];


/**
 * One line in a group
 *
 * @since 1.8.10
 *
 * @param line The single line of conditional rule.
 * @param {} strings Translation field
 * @param boolean isFirst
 * @param formFields
 * @param onRemoveLine
 * @param onUpdateLine
 * @returns {*}
 * @constructor
 */
export const ConditionalLine = ({line, strings, isFirst, formFields, onRemoveLine, onUpdateLine}) => {
    const {value, field, compare} = line;
    const lineId = line.id;
    const fieldConfig = React.useMemo(() => {
        const config = formFields.find(f => f.ID === field);
        if (config) {
            return config;
        }
        return {
            type: 'text'
        }
    }, [field]);

    /**
     * Callback for changing which field this line is based on
     *
     * @param field
     */
    const onChangeField = (field) => {
        onUpdateLine({
            field, compare, value,
        })
    };

    /**
     * Callback for changing a line's value
     *
     * @since 1.8.10
     *
     * @param value
     */
    const onChangeValue = (value) => {
        onUpdateLine({
            field, compare, value,
        })
    };

    /**
     * Callback for changing a line's comparison type
     *
     * @since 1.8.10
     *
     * @param compare
     */
    const onChangeCompare = (compare) => {
        onUpdateLine({
            field, compare, value,
        })
    };


    return (
        <div key={lineId} className={`caldera-condition-line condition-line-${lineId}`}>
            <span style={{display: "inline-block"}}>
                {isFirst ? <React.Fragment>{strings['if']}</React.Fragment> :
                    <React.Fragment>{strings['and']}</React.Fragment>}
            </span>
            <select
                style={{maxWidth: "120px", verticalAlign: 'inherit'}}
                className="condition-line-field"
                value={field}
                onChange={(e) => onChangeField(e.target.value,lineId)}
            >
                <option/>
                <optgroup label={strings['fields']}>
                    {formFields.map(field => (
                        <option value={field.ID}>
                            {field.label} [{field.slug}]
                        </option>
                    ))}
                </optgroup>
            </select>
            <select
                className="condition-line-compare"
                value={compare}
                onChange={(e) => onChangeCompare(e.target.value)}
                style={{maxWidth: "120px", verticalAlign: 'inherit'}}
            >
                {['is', 'isnot', 'greater', 'smaller', 'startswith', 'endswith', 'contains'].map(compareType => (
                    <option key={compareType} value={compareType}>{compareType}</option>
                ))}
            </select>
            {isFieldTypeWithOptions(fieldConfig.type) ? (
                <select
                    value={value}
                    onChange={(e) => onChangeValue(e.target.value)}
                    style={{maxWidth: "165px", verticalAlign: 'inherit'}}>
                    <option/>
                    {formFields.map(option => {
                        //should be fieldConfig.option, except that doesn't exist.
                        return (
                            <option
                                key={option.value}
                                value={option.value}
                            >
                                {option.label}
                            </option>
                        )
                    })}
                </select>
            ) : (
                <input
                    onChange={onChangeValue}
                    value={value}
                    type="text"
                    className="magic-tag-enabled block-input"
                />
            )}
            <button
                onClick={(e) => {
                    e.preventDefault();
                    onRemoveLine(lineId);
                }}
                className="button pull-right" type="button"
            >
                <i className="icon-join"/>
            </button>
        </div>
    );
};

/**
 * All of the lines of one conditional
 *
 * @since 1.8.10
 *
 * @param lines
 * @param strings
 * @param formFields
 * @param onRemoveLine
 * @param onUpdateLine
 * @returns {*}
 * @constructor
 */
const ConditionalLines = ({lines, strings, formFields, onRemoveLine, onUpdateLine}) => {
    return (
        <div className="caldera-condition-group caldera-condition-lines">
            {lines.map((line, index) => {
                return (
                    <ConditionalLine
                        line={line}
                        isFirst={0 === index}
                        onUpdateLine={(update)=> onUpdateLine(update,line.id)}
                        {...{
                            strings,
                            formFields,
                            onRemoveLine,
                        }}
                    />
                )
            })}

            <div style={{margin: "12px 0 0"}}>
                <button
                    className="button button-small condition-group-add-line"
                    type="button"
                >
                    {strings['add-condition']}
                </button>
            </div>
        </div>)
};

/**
 * One conditional
 *
 * @since 1.8.10
 */
const Conditional = ({conditional, formFields, strings, onRemoveConditional, onUpdateConditional, fieldsNotAllowed, fieldsUsed}) => {
    const {type, config, id} = conditional;
    const group = config && config.hasOwnProperty('group') ? config.group : {};
    const name = conditional.hasOwnProperty('config') && conditional.config.hasOwnProperty('name') ? conditional.config.name : '';
    const onAddLine = (groupId) => {
        const id = `cl_${Math.random().toString(36).replace(/[^a-z]+/g, '').substr(0, 5)}`
        onUpdateConditional({
            conditional,
            config: {
                ...conditional.config,
                group: {
                    ...group,
                    [groupId] : {
                        ...conditional.config[groupId],
                        [id]: {
                            field: '',
                            value: '',
                            type: ''
                        }
                    }

                }
            }
        });
    };

    /**
     * Callback for updating a line in a conditional group
     *
     * @since 1.8.10
     *
     * @param line
     * @param lineId
     */
    const onUpdateLine = (line, lineId,groupId) => {
        let group = conditional.config.group;
        onUpdateConditional({
            ...conditional,
            config: {
                ...conditional.config,
                group: {
                    ...group,
                    [groupId] :{
                        ...group[groupId],
                        [lineId]: line
                    }
                }
            }
        });
    };

    /**
     *  Callback for removing a line from a line group
     *  
     *  @since 1.8.10
     *
     * @param lineId
     * @param groupId
     */
    const onRemoveLine = (lineId, groupId) => {
        let groups = group;

        //Delete line from line group
        delete groups[groupId][lineId];

        //no lines left?
        if( ! Object.values(groups[groupId]).length){
            //delete line group
            delete groups[groupId];
        }
        onUpdateConditional({
            ...conditional,
            config: {
                ...conditional.config,
                group: groups
            }
        });
    };


    /**
     * Callback for updating name of conditional group
     *
     * @since 1.8.10
     *
     * @param name
     */
    const onUpdateName = (name) => {
        onUpdateConditional({
            ...conditional,
            config: {
                ...conditional.config,
                name,
            }
        })
    };

    /**
     * Callback for changing type of conditional
     *
     * @since 1.8.10
     *
     * @param type
     * @returns {*}
     */
    const onChangeType = (type) => {
        return onUpdateConditional({
            ...conditional,
            type
        })
    };


    return (
        <div className="caldera-editor-condition-config caldera-forms-condition-edit"
             style={{marginTop: '-27px', width: "auto"}}
        >

            <div
                className={`condition-point-${id}`}
                style={{float: 'left', width: "550px"}}
            >
                <div
                    className="caldera-config-group"
                >
                    <label htmlFor={`condition-group-name-${id}`}>
                        {strings.name}
                    </label>
                    <div
                        className="caldera-config-field"
                    >
                        <input
                            type="text"
                            name={`conditions[${id}][name]`}
                            id={`condition-group-name-${id}`}
                            value={name}
                            onChange={(e) => onUpdateName(e.target.value)}
                            required
                            className="required block-input condition-group-name"
                        />
                    </div>
                </div>

                <div className="caldera-config-group">
                    <label
                        htmlFor={`condition-group-type-${id}`}
                    >
                        {strings.type}
                    </label>
                    <div

                        className="caldera-config-field"
                    >
                        <select
                            value={type}
                            onChange={(e) => onChangeType(e.target.value)}
                            id={`condition-group-type-${id}`}
                            name={`conditions[${id}][type]`}
                            className="condition-group-type"
                        >
                            <option value=""/>
                            <option value="show">
                                {strings.show}
                            </option>
                            <option value="hide">
                                {strings.hide}
                            </option>
                            <option value="disable">
                                {strings.disable}
                            </option>
                        </select>
                        {type &&
                        <button
                            type="button"
                            className="pull-right button button-small condition-group-add-lines"
                            onClick={(e) => {
                                e.preventDefault();
                                onAddLine(groupId)
                            }}
                        >
                            {strings['add-conditional-line']}
                        </button>
                        }
                    </div>
                </div>
                {Object.keys(group).map(groupId => {
                    const lineGroups = group[groupId];
                    const lines = Object.keys(lineGroups).map(
                        (lineGroupId) => {
                            return {
                                ...lineGroups[lineGroupId],
                                id: lineGroupId,
                            }
                        }
                    );
                    return (
                        <Fragment>
                            <ConditionalLines
                                lines={lines}
                                strings={strings}
                                formFields={formFields}
                                onUpdateLine={(update,lineId) => onUpdateLine(update,lineId,groupId)}
                                onRemoveLine={(lineId) => onRemoveLine(lineId,groupId)}
                            />
                        </Fragment>
                    )
                })

                }
                <button
                    style={{margin: "12px 0 12px"}}
                    type="button"
                    className="block-input button"
                    data-confirm={strings['confirm-remove']}
                    onClick={e => {
                        e.preventDefault();
                        onRemoveConditional(id)
                    }}
                >
                    {strings['remove-condition']}
                </button>

                <AppliesToFields
                    formFields={formFields}
                    fieldsUsed={fieldsUsed}
                    fieldsNotAllowed={fieldsNotAllowed}
                    strings={strings}
                    onChange={() => {
                    }}
                    groupId={id}
                />
            </div>

        </div>
    )
};


export default Conditional;