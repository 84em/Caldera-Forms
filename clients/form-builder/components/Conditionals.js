import React from 'react';
import cfEditorState from "@calderajs/cf-editor-state";

/**
 * One item in the conditionals list
 *
 * @since 1.8.10
 *
 * @param active
 * @param id
 * @param name
 * @returns {*}
 * @constructor
 */
const ConditionalListItem = ({active, id, name, onChooseItem}) => {
    return (
        <li className={`caldera-condition-nav ${active} caldera-forms-condition-group condition-point-${id}`}>
            <a className="condition-open-group" onClick={(e) => {
                e.preventDefault();
                onChooseItem(id);
            }}
               style={{cursor: "pointer"}}
            >
            <span id={`condition-group-${id}`}>
                {name}
            </span>
                <span className="condition-line-number"/>
            </a>
        </li>
    );
};

/**
 * The list of conditionals
 *
 * @since 1.8.10
 *
 * @param conditionals
 * @returns {*}
 * @constructor
 */
export const ConditionalsList = ({conditionals, onChooseItem}) => {
    return (
        <div className="caldera-editor-conditions-panel" style={{marginBottom: "32px"}}>
            <ul className="active-conditions-list">
                {conditionals.map(condition => <ConditionalListItem key={condition.id} {...condition}
                                                                    onChooseItem={onChooseItem}/>)}
            </ul>
        </div>
    )
};

/**
 * Set name for a group
 *
 * @since 1.8.10
 *
 * @param placeholder
 * @param onChange
 * @param id
 * @returns {*}
 * @constructor
 */
export const NewGroupName = ({placeholder, onChange, id}) => {
    return (
        <input type="text"
               name={`conditions[${id}][name]`}
               value="{{name}}"
               className="condition-new-group-name"
               placeholder={placeholder}
               style={{width: "100%"}}
        />
    )
};

/**
 * The section to set which fields it applies to
 *
 * @since 1.8.10
 *
 * @param fields
 * @param fieldsUsed
 * @param strings
 * @returns {*}
 * @constructor
 */
const AppliesToFields = ({fields, fieldsUsed, appliedFields, strings, onChange, groupId}) => {

    return (
        <div style={{float: 'left', width: '288px', paddingLeft: '12px'}}>
            <h4 style={{borderBottom: `1px solid rgb(191, 191, 191)`, margin: `0px 0px 6px; padding: 0px 0px 6px`}}>
                {strings['applied-fields']}
            </h4>
            <p className="description">{strings['select-apply-fields']}</p>
            {fields.map(field => (
                    <label style={{display: 'block', marginLeft: `20px`}}>
                        {field.label}
                        <input
                            style={{marginLeft: `-20px`}}
                            type="checkbox"
                            disabled={fieldsUsed.includes(field.id)}
                            onClick={(e) => {
                                e.preventDefault();
                                if (!appliedFields || !appliedFields.length) {
                                    onChange([field.id])

                                } else if (appliedFields.include(field.id)) {
                                    onChange(appliedFields.filter(f => f.id === field.id))
                                } else {
                                    onChange([...appliedFields, field.id])
                                }
                            }}
                            value={groupId}
                        />
                    </label>
                )
            )}
        </div>
    );
};

/**
 * One line in a group
 *
 * @since 1.8.10
 *
 * @param line
 * @param strings
 * @param isFirst
 * @param fieldOptions
 * @returns {*}
 * @constructor
 */
const ConditionalLine = ({line, strings, isFirst, fieldOptions}) => {

    return (
        <div key={line.id} className={`caldera-condition-line condition-line-${line.id}`}>
                        <span style={{display: "inline-block"}}>
                            {isFirst ? <React.Fragment>{strings['if']}</React.Fragment> :
                                <React.Fragment>{strings['and']}</React.Fragment>}
                        </span>

            <select
                style={{maxWidth: "120px", verticalAlign: 'inherit'}}
                className="condition-line-field"
            >
                <option/>
                <optgroup label={strings['fields']}>
                    {fields.map(field => (
                        <option value={field.id}>
                            {field.label} [{field.slug}]
                        </option>
                    ))}
                </optgroup>
            </select>
            <select
                className="condition-line-compare"
                style={{maxWidth: "120px", verticalAlign: 'inherit'}}
            >
                {['is', 'isnot', 'greater', 'smaller', 'startswith', 'endswith', 'contains'].map(compareType => (
                    <option key={compareType} value={compareType}>{compareType}</option>
                ))}

            </select>
            {fieldOptions ? (<select
                    style={{maxWidth: "165px", verticalAlign: 'inherit'}}>
                    <option/>
                    {fieldOptions.map(option => <option key={option.value}
                                                        value={option.value}>{option.label}</option>)}

                </select>
            ) : (
                <input
                    type="text"
                    className="magic-tag-enabled block-input"
                />
            )}
        </div>
    );
};

/**
 * All of the lines of one conditional
 *
 * @since 1.8.10
 *
 * @param lines
 * @param strings
 * @returns {*}
 * @constructor
 */
const ConditionalLines = ({lines, strings}) => {

    let isFirst = true;
    return (
        <div className="caldera-condition-group caldera-condition-lines">
            {lines.map(line => {
                const r = (
                    <ConditionalLine line={line} isFirst={isFirst} strings={strings}/>
                );
                isFirst = false;
                return r;
            })}

            <div style={{margin: "12px 0 0"}}>
                <button
                    className="button button-small condition-group-add-line"
                    type="button"
                >
                    {strings['add-condition']}
                </button>
            </div>
        </div>)
};

/**
 * One conditional
 *
 * @since 1.8.10
 *
 * @param name
 * @param strings
 * @param id
 * @param type
 * @param group
 * @param fields
 * @returns {*}
 * @constructor
 */
const Conditional = ({name, strings, id, type, group, fields}) => {
    const fieldsUsed = React.useMemo(() => {
        return fields;
    }, [id]);
    const onAddLine = () => {
    };
    return (
        <div className="caldera-editor-condition-config caldera-forms-condition-edit"
             style={{marginTop: '-27px', width: "auto"}}
        >
            {name ? (
                <NewGroupName placeholder={strings['new-group-name']} id={id}/>
            ) : (
                <div
                    className={`condition-point-${id}`}
                    style={{float: 'left', width: "550px"}}
                >

                    <div className="caldera-config-group">
                        <label htmlFor={`condition-group-name-${id}`}>
                            {strings.name}
                        </label>
                        <div className="caldera-config-field">
                            <input
                                type="text"
                                name={`conditions[${id}][name]`}
                                id={`condition-group-name-${id}`}
                                value="{{name}}"
                                required
                                className="required block-input condition-group-name"
                            />
                        </div>
                    </div>

                    <div className="caldera-config-group">
                        <label htmlFor={`condition-group-type-${id}`}>
                            {strings.type}
                        </label>
                        <div className="caldera-config-field">
                            <select
                                id={`condition-group-type-${id}`}
                                name={`conditions[${id}][type]`}
                                className="condition-group-type"
                            >
                                <option value=""/>
                                <option value="show">
                                    {strings.show}
                                </option>
                                <option value="hide">
                                    {strings.hide}
                                </option>
                                <option value="disable">
                                    {strings.disable}
                                </option>
                            </select>
                            {type &&
                            <button
                                type="button"
                                className="pull-right button button-small condition-group-add-lines"
                                onClick={onAddLine}
                            >
                                {strings['add-conditional-line']}
                            </button>
                            }
                        </div>
                    </div>
                    <ConditionalLines lines={group.lines} strings={strings}/>
                    <button
                        style={{margin: "12px 0 12px"}}
                        type="button"
                        className="block-input button"
                        data-confirm={strings['confirm-remove']}
                    >
                        {strings['remove-condition']}
                    </button>
                    <AppliesToFields fields={fields} fieldsUsed={fieldsUsed}/>
                </div>
            )}
        </div>
    )
};

/**
 * A button to add a new conditional with
 *
 * @param text
 * @param onClick
 * @returns {*}
 * @constructor
 */
export const NewConditionalButton = ({text, onClick}) => {
    return (
        <button
            style={{width: "250px"}}
            id="new-conditional"
            className="button"
            type="button"
            onClick={e => {
                e.preventDefault();
                onClick();
            }}
        >
            {text}
        </button>
    )
};


/**
 *
 * @param {cfEditorState }state
 * @param strings
 * @returns {*}
 */
export default function ({state, strings, fields}) {
    /**
     * Ref for state object
     *
     * @type {React.MutableRefObject<cfEditorState>}
     */
    const editorState = React.useRef(state);

    /**
     * Tracks the open conditional
     */
    const [openCondition, setOpenCondition] = React.useState('');

    /**
     * The conditional group that is open
     *
     * @type {function(): conditional}
     */
    const getOpenConditional = React.useCallback(() => {
            return state.getConditional(openCondition);
        }, [openCondition]
    );

    /**
     * Callback for adding conditional
     * @param name
     */
    const onNewConditional = (name) => {
        editorState.current.addConditional({type: 'show', name});
        setOpenCondition(name);
    };

    /**
     * Callback for adding conditional
     * @param name
     */
    const onAddConditional = (conditionalId) => {
        editorState.current.removeConditional(conditionalId);
        setOpenCondition('');
    };

    /**
     * Callback for removing conditional
     * @param name
     */
    const onRemoveConditional = (conditionalId) => {
        editorState.current.removeConditional(conditionalId);
        setOpenCondition('');
    };

    return (
        <React.Fragment>
            <NewConditionalButton
                text={strings['new-conditional']}
                onClick={onNewConditional}
            />
            <ConditionalsList
                conditionals={editorState.current.getAllConditionals()}
                onChooseItem={setOpenCondition}
            />
            {openCondition &&
            <Conditional
                field={fields}
                group={getOpenConditional()}
                fields={fields}
                id={openCondition}
                strings={strings}
                onAddConditional={onAddConditional}
                onRemoveConditional={onRemoveConditional}
            />
            }
        </React.Fragment>
    )
}
