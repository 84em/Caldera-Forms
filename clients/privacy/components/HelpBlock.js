import {createElement} from "@wordpress/element";

export const HelpBlock = ({className, children}) =>
    (
        <div
            className={className ? className : 'screen-reader-text'}
        >
            {children}
        </div>
    );
