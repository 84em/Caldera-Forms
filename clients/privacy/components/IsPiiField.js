import React from  'react';
import PropTypes from 'prop-types';
import { FormGroup,FormLabel,Form,FormCheck } from 'react-bootstrap';
const{Text} = Form;
/**
 * Determine if field is a PII field
 *
 * @since 1.7.0
 *
 * @param {Object} field
 * @param {Object} privacySettings
 * @returns {*}
 */
function fieldIsPii(field,privacySettings ) {
    return privacySettings.piiFields.length && privacySettings.piiFields.includes(field.ID);
}

/**
 * Setting to determine if this is a PII field
 *
 * @since 1.7.0
 *
 * @param props
 * @returns {*}
 * @constructor
 */
export const IsPiiField = (props) => {
    const idAttr  = `caldera-forms-privacy-gdpr-is-pii-field-${props.field.ID}`;
    return (
        <FormGroup>
            <FormLabel
               htmlFor={idAttr}
            >
               Personally Identifying Field?
            </FormLabel>

            <Form.Check
                type={'checkbox'}
                label={'Yes'}
                onChange={() => {
                    props.onCheck(props.field.ID)
                }}
                checked={fieldIsPii(props.field, props.privacySettings)}
            />

            <Text
                className={'screen-reader-text'}
            >
                Does field contain personally identifying data?
            </Text>
        </FormGroup>
    );

};

IsPiiField.propTypes = {
    field: PropTypes.object.isRequired,
    privacySettings: PropTypes.object.isRequired,
    onCheck: PropTypes.func.isRequired,
};