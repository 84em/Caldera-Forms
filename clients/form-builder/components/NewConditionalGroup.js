import React, {Fragment} from "react";
import {NewConditionalButton} from "./Conditionals";

/**
 * UI for adding a conditional group
 *
 * @since 1.8.10
 *
 * @param strings
 * @param onNewConditional
 * @returns {*}
 * @constructor
 */
export const NewConditionalGroup = ({strings, onNewConditional}) => {
    //Tracks the new group's name
    const [newGroupName, setNewGroupName] = React.useState('');

    //Tracks the new group's ID
    const [newGroupId, setNewGroupId] = React.useState('');

    //Should we show new group name input?
    const [showNewGroupName, setShowNewGroupName] = React.useState(false);

    //Ref for new group name input so we can control focus.
    let newNameInputRef = React.useRef('');

    /**
     * When new group name input is blurred:
     * * If has name, add group and hide new group name.
     * * If no name, hide new group name.
     *
     * @since 1.8.10
     *
     */
    const onNewBlur = () => {
        if (newGroupName.length) {
            onNewConditional(newGroupName, newGroupId);
            setShowNewGroupName(false);
        } else {
            setShowNewGroupName(false);
            setNewGroupName('');
            setNewGroupId('');
        }
    };

    /**
     * When we click on the new group button:
     * * Create ID for group
     * * Show new group name input
     * * Focus new group name input
     *
     * @since 1.8.10
     */
    const onClickNew = () => {
        const id = `con_${Math.random().toString(36).replace(/[^a-z]+/g, '').substr(0, 5)}`;
        setNewGroupId(id);
        setShowNewGroupName(true);
        setNewGroupName('');
        newNameInputRef.current.focus();
    };

    return (
        <Fragment>
            {!showNewGroupName && <NewConditionalButton
                text={strings['new-conditional']}
                onClick={onClickNew}
            />}

            <input
                ref={newNameInputRef}
                type="text"
                name={`conditions[${newGroupId}][name]`}
                value={newGroupName}
                className="condition-new-group-name"
                placeholder={strings['new-group-name']}
                style={{width: "100%",display:showNewGroupName ? 'block' : 'none'}}
                onChange={(e) => setNewGroupName(e.target.value)}
                onBlur={onNewBlur}
            />

        </Fragment>
    )

}