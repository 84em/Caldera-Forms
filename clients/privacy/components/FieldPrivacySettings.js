import React from 'react';
import PropTypes from 'prop-types';
import {Col,Row} from 'react-bootstrap';
import {IsEmailIdentifyingField} from "./IsEmailIdentifyingField";
import {IsPiiField} from "./IsPiiField";

/**
 * Privacy Settings for one field of a form
 *
 * @since 1.7.0
 *
 * @param props
 * @returns {*}
 * @constructor
 */
export const FieldPrivacySettings = (props) => {
    return (
        <Row
            style={{
                backgroundColor: '#fff',
                margin: '1em',
            }}
        >
            <h4
                style={{
                    margin: 0,
                    padding: '.7em 1em',
                    borderBottom: '1px solid #eee'
                }}
            >
                {props.field.name}
            </h4>
            <Row
                xs={6}
                style={{
                    padding: '12px'
                }}
            >
                <IsPiiField
                    field={props.field}
                    privacySettings={props.privacySettings}
                    onCheck={props.onCheckIsPii}
                />
            </Row>

            <Row
                xs={6}
                style={{
                    padding: '12px'
                }}
            >
                <IsEmailIdentifyingField
                    className={'col-xs-6'}
                    field={props.field}
                    privacySettings={props.privacySettings}
                    onCheck={props.onCheckIsEmail}
                />
            </Row>


        </Row>
    );
};

FieldPrivacySettings.propTypes = {
    field: PropTypes.object.isRequired,
    formId: PropTypes.string.isRequired,
    privacySettings: PropTypes.object.isRequired,
    onCheckIsEmail: PropTypes.func.isRequired,
    onCheckIsPii: PropTypes.func.isRequired
};