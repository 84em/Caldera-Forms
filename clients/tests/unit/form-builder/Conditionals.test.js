import React from 'react';
import {ConditionalsList,NewConditionalButton,NewGroupName} from "../../../form-builder/components/Conditionals";
import testForm from './test-form';
import renderer from 'react-test-renderer';


describe('Conditionals', () => {
    it('Renders list', () => {
        const component = renderer.create(
            <ConditionalsList
                conditionals={Object.values(testForm.conditional_groups.conditions)}
            />
        );
        expect(component.toJSON()).toMatchSnapshot();
    });

    it( 'Renders new conditional button', () => {
        const component = renderer.create(
            <NewConditionalButton
                text={'New'}
                onClick={jest.fn()}
            />
        );
        expect(component.toJSON()).toMatchSnapshot();
    });

    it( 'New group name field renders',() => {
        const props = {placeholder: 'Placeholder', onChange: jest.fn(), id: 'f', value: 'val'};
        const component = renderer.create(
            <NewGroupName
                text={'New'}
                onClick={jest.fn()}
            />
        );
        expect(component.toJSON()).toMatchSnapshot();
    });

});