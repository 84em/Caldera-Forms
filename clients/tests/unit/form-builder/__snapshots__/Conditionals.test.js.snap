// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
<div
  className="caldera-editor-conditions-panel"
  style={
    Object {
      "marginBottom": "32px",
    }
  }
>
  <ul
    className="active-conditions-list"
  >
    <li
      className="caldera-condition-nav undefined caldera-forms-condition-group condition-point-con_3840644613791445"
    >
      <a
        className="condition-open-group"
        data-open-group="con_3840644613791445"
        style={
          Object {
            "cursor": "pointer",
          }
        }
      >
        <span
          id="condition-group-con_3840644613791445"
        >
          Hide Number
        </span>
        <span
          className="condition-line-number"
        />
      </a>
    </li>
    <li
      className="caldera-condition-nav undefined caldera-forms-condition-group condition-point-con_3156693554561454"
    >
      <a
        className="condition-open-group"
        data-open-group="con_3156693554561454"
        style={
          Object {
            "cursor": "pointer",
          }
        }
      >
        <span
          id="condition-group-con_3156693554561454"
        >
          Hide Dropdown
        </span>
        <span
          className="condition-line-number"
        />
      </a>
    </li>
    <li
      className="caldera-condition-nav undefined caldera-forms-condition-group condition-point-con_8761120514939434"
    >
      <a
        className="condition-open-group"
        data-open-group="con_8761120514939434"
        style={
          Object {
            "cursor": "pointer",
          }
        }
      >
        <span
          id="condition-group-con_8761120514939434"
        >
          Hide Number 2
        </span>
        <span
          className="condition-line-number"
        />
      </a>
    </li>
  </ul>
</div>
`;

exports[`Conditionals Renders list 1`] = `
<div
  className="caldera-editor-conditions-panel"
  style={
    Object {
      "marginBottom": "32px",
    }
  }
>
  <ul
    className="active-conditions-list"
  >
    <li
      className="caldera-condition-nav undefined caldera-forms-condition-group condition-point-con_3840644613791445"
    >
      <a
        className="condition-open-group"
        data-open-group="con_3840644613791445"
        style={
          Object {
            "cursor": "pointer",
          }
        }
      >
        <span
          id="condition-group-con_3840644613791445"
        >
          Hide Number
        </span>
        <span
          className="condition-line-number"
        />
      </a>
    </li>
    <li
      className="caldera-condition-nav undefined caldera-forms-condition-group condition-point-con_3156693554561454"
    >
      <a
        className="condition-open-group"
        data-open-group="con_3156693554561454"
        style={
          Object {
            "cursor": "pointer",
          }
        }
      >
        <span
          id="condition-group-con_3156693554561454"
        >
          Hide Dropdown
        </span>
        <span
          className="condition-line-number"
        />
      </a>
    </li>
    <li
      className="caldera-condition-nav undefined caldera-forms-condition-group condition-point-con_8761120514939434"
    >
      <a
        className="condition-open-group"
        data-open-group="con_8761120514939434"
        style={
          Object {
            "cursor": "pointer",
          }
        }
      >
        <span
          id="condition-group-con_8761120514939434"
        >
          Hide Number 2
        </span>
        <span
          className="condition-line-number"
        />
      </a>
    </li>
  </ul>
</div>
`;

exports[`Conditionals Renders new conditional button 1`] = `
<button
  className="button"
  id="new-conditional"
  onClick={[MockFunction]}
  style={
    Object {
      "width": "250px",
    }
  }
  type="button"
>
  New
</button>
`;
